jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+
+std::unordered_map<string, faiss::MetricType> mapMetric = {
+        {"l2", faiss::METRIC_L2},
+        {"innerproduct", faiss::METRIC_INNER_PRODUCT}
+};
+
+extern "C"
+
+struct JavaException {
+    JavaException(JNIEnv* env, const char* type = "", const char* message = "")
+    {
+        jclass newExcCls = env->FindClass(type);
+        if (newExcCls != NULL)
+            env->ThrowNew(newExcCls, message);
+    }
+};
+
+inline void has_exception_in_stack(JNIEnv* env)
+{
+    if (env->ExceptionCheck() == JNI_TRUE)
+        throw std::runtime_error("Exception Occured");
+}
+
+void catch_cpp_exception_and_throw_java(JNIEnv* env)
+{
+    try {
+        throw;
+    }
+    catch (const std::bad_alloc& rhs) {
+        JavaException(env, "java/io/IOException", rhs.what());
+    }
+    catch (const std::runtime_error& re) {
+        JavaException(env, "java/lang/Exception", re.what());
+    }
+    catch (const std::exception& e) {
+        JavaException(env, "java/lang/Exception", e.what());
+    }
+    catch (...) {
+        JavaException(env, "java/lang/Exception", "Unknown exception occured");
+    }
+}
+
+/**
+ * Method: saveIndex
+ *
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex_saveIndex
+(JNIEnv* env, jclass cls, jintArray ids, jobjectArray vectors, jstring indexPath, jobjectArray algoParams, jstring spaceType)
+{
+	vector<int64_t> idVector;
+	vector<float>   dataset;
+	vector<string> paramsList;
+	string indexDescription = "HNSW32";
+	faiss::MetricType metric = faiss::METRIC_L2;
+	std::unique_ptr<faiss::Index> indexWriter;
+	int dim = 0;
+	try {
+		//---- ids
+		int* object_ids = NULL;
+		object_ids = env->GetIntArrayElements(ids, 0);
+		for(int i = 0; i < env->GetArrayLength(ids); ++i) {
+			idVector.push_back(object_ids[i]);
+		}
+		env->ReleaseIntArrayElements(ids, object_ids, 0);
+        has_exception_in_stack(env);
+
+		//---- vectors
+		for (int i = 0; i < env->GetArrayLength(vectors); ++i) {
+			jfloatArray vectorArray = (jfloatArray)env->GetObjectArrayElement(vectors, i);
+			float* vector = env->GetFloatArrayElements(vectorArray, 0);
+			dim = env->GetArrayLength(vectorArray);
+			for(int j = 0; j < dim; ++j) {
+				dataset.push_back(vector[j]);
+			}
+			env->ReleaseFloatArrayElements(vectorArray, vector, 0);
+		}
+		has_exception_in_stack(env);
+
+		//---- indexPath
+		const char *indexString = env->GetStringUTFChars(indexPath, 0);
+        string indexPathString(indexString);
+        env->ReleaseStringUTFChars(indexPath, indexString);
+        has_exception_in_stack(env);
+
+		//---- algoParams
+		int paramsCount = env->GetArrayLength(algoParams);
+        for (int i=0; i<paramsCount; i++) {
+            jstring param = (jstring) (env->GetObjectArrayElement(algoParams, i));
+            const char *rawString = env->GetStringUTFChars(param, 0);
+            paramsList.push_back(rawString);
+
+            int M = 32;
+            if (sscanf(rawString, "M=%d", &M) == 1) {
```

Comment:
Why do we need to sscanf here?

Reply:
> Why do we need to sscanf here?

In the KNN plugin we would make algoParam like: `M=32, efSearch=20`, As [faiss index factory docs](https://github.com/facebookresearch/faiss/wiki/The-index-factory#non-exhaustive-search-components) shows, if `M=32` we need use string: `HNSW32` for index Description. so i use sscanf to get M.

i think we need an adapter to match different knnEngine algorithms params. if you have good idea, pls let me know.

Reply:
Ah got it. I think makes sense.